### DISCUSSION TITLE
How about making podman machine start implicit?

### DISCUSSION THREAD

**Discussion by ssbarnea** [containers/podman#14289] [Q&A]
Every time I restart macos I discover that some podman commands fail because the podman machine was not started.

As configuring automatic starting at boot time is complex on macos, I would like to propose making the `podman machine start` implicit when podman is executed.

## Current behavior
```
Running command:
  podman build -f ./Containerfile -t quay.io/ansible/creator-ee:latest .
Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman. failed to create sshClient: connection to bastion host (ssh://core@localhost:50691/run/user/501/podman/podman.sock) failed: dial tcp [::1]:50691: connect: connection refused

An error occured (rc=125), see output line(s) above for details.
```

## Desired behavior

When a command that requires a server running is called, first podman should attempt to run either `podman machine init` or `podman machine start` before failing.

Likely that in 99% of cases, that is what user needed, so better to attempt to do it instead of asking him to do it.

That feature is particularly useful on macos, where the machine is needed.

Related: #11541

**Comment by rhatdan** [comment]
@baude @jwhonce @ashley-cui, Do we have the information to make this happen.  If the primary connection is to a local port, and the connection does not exists, should we check if podman machine is started, if not start it, sleep and then try the connection again?

**Reply by baude** [reply]
I have zero idea there.  Need our mac folks to chime in.

**Reply by ssbarnea** [reply]
Now I that my main machine is a M1 one, you can rely on my support. Happy to test and provide support for anything related to arm64, not only on MacOS, but also on Fedora/Windows, as I have the entire trinity running on arm64.

Sadly I might not be able to make the code changes myself because I know no `go`.

I should also mention that I am really pleased about the experience of downloading the machine(VM) on mac, it was really fast and simple.

If we do have the place in the code to hook the on-demand start feature it would be awesome to do it. Maybe we could only make it an opt-in that is activated by defining and environment variable until it matures and we can just make it enabled by default.

Ability to use podman is key for my team because one of our products is vscode-ansible extension, which mostly relies on running a container engine.

**Comment by afbjorklund** [comment]
You get the same output, even when **not** using `podman machine` at all. So it is not always helpful.

```
Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman socket: Get "http://d/v3.4.2/libpod/_ping": dial unix /home/anders/.lima/podman/sock/podman.sock: connect: no such file or directory
```

The docker output makes less assumptions:

```
Cannot connect to the Docker daemon at unix:///home/anders/.lima/docker/sock/docker.sock. Is the docker daemon running?
```

----

Now both of them are supported out-of-the-box, in addition to the default `nerdctl` (i.e. for containerd)

```
brew install lima
limactl start template://podman
# later: `limactl start podman`
podman.lima run ....
```

So very similar to:

```
brew install lima
limactl start template://docker
# later: `limactl start docker`
docker.lima run ....
```

**Reply by afbjorklund** [reply]
Or at least they _will_ be, after the next release. Currently the procedure is a bit longer, as shown in today's presentation:

https://sched.co/ytpi _"Running Containerd and k3s on MacOS - Akihiro Suda, NTT Corporation & Jan Dubois, SUSE"_

```console
$ limactl start template://podman
$ brew install podman
$ podman system connection add lima ⧵
"unix://$HOME/.lima/podman/sock/podman.sock"
$ podman system connection default lima
$ podman run ...
```

```console
$ limactl start template://docker
$ brew install docker
$ docker context create lima --docker ⧵
"host=unix://$HOME/.lima/docker/sock/docker.sock"
$ docker context use lima
$ docker run ...
```

Currently it fails if you try to install the podman client from brew, since it is not compatible with the server (running in Ubuntu).

```
Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman socket: server API version is too old. Client "4.0.0" server "3.4.4"
```

So you need to use `limactl shell podman podman`, instead. This will avoid the client/server API, and just run the command.
